core/finality-grandpa/src/communication/gossip.rs

```diff
@@ -593,6 +663,129 @@ impl<Block: BlockT> Inner<Block> {
 		Action::ProcessAndDiscard(topic, benefit::BASIC_VALIDATED_COMMIT)
 	}
 
+	fn validate_catch_up_message(&mut self, who: &PeerId, full: &FullCatchUpMessage<Block>)
+		-> Action<Block::Hash>
+	{
+		match &self.pending_catch_up {
+			PendingCatchUp::Requesting { who: peer, request, instant } => {
+				if peer != who {
+					return Action::Discard(Misbehavior::OutOfScopeMessage.cost());
+				}
+
+				if request.set_id != full.set_id {
+					return Action::Discard(cost::MALFORMED_CATCH_UP);
+				}
+
+				if request.round.0 > full.message.round_number {
+					return Action::Discard(cost::MALFORMED_CATCH_UP);
+				}
+
+				if full.message.prevotes.is_empty() || full.message.precommits.is_empty() {
+					return Action::Discard(cost::MALFORMED_CATCH_UP);
+				}
+
+				// move request to pending processing state, we won't push out
+				// any catch up requests until we import this one (either with a
+				// success or failure).
+				self.pending_catch_up = PendingCatchUp::Processing {
+					instant: instant.clone(),
+				};
+
+				// always discard catch up messages, they're point-to-point
+				let topic = super::global_topic::<Block>(full.set_id.0);
+				Action::ProcessAndDiscard(topic, benefit::BASIC_VALIDATED_CATCH_UP)
+			},
+			_ => Action::Discard(Misbehavior::OutOfScopeMessage.cost()),
+		}
+	}
+
+	fn note_catch_up_message_processed(&mut self) {
+		match &self.pending_catch_up {
+			PendingCatchUp::Processing { .. } => {
+				self.pending_catch_up = PendingCatchUp::None;
+			},
+			state => trace!(target: "afg",
+				"Noted processed catch up message when state was: {:?}",
+				state,
+			),
+		}
+	}
+
+	fn handle_catch_up_request(
+		&mut self,
+		who: &PeerId,
+		request: CatchUpRequestMessage,
+		set_state: &environment::SharedVoterSetState<Block>,
+	) -> (Option<GossipMessage<Block>>, Action<Block::Hash>) {
+		let local_view = match self.local_view {
+			None => return (None, Action::Discard(Misbehavior::OutOfScopeMessage.cost())),
+			Some(ref view) => view,
+		};
+
+		if request.set_id != local_view.set_id {
+			return (None, Action::Discard(Misbehavior::OutOfScopeMessage.cost()));
```

Comment:
There is unfortunately a race here, when sending a catch-up request right at the point your peer changes. Could we have a bit of leeway, where if we're within `CATCH_UP_THRESHOLD` of the beginning of the new set, we don't charge a high cost?

(that's also still a bit of a race, since we might move to a new set and then catch up ourselves)

Reply:
Implemented your suggestion.